### DISCUSSION TITLE
Use per-file-ignores with Ruff extension for Zed

### DISCUSSION THREAD

**Discussion by prepin** [astral-sh/ruff#14642] [Q&A] [answered]
I'm unsure whether this question belongs here or on the Zed discussion board, but I'll ask here for now.

I want to configure Ruff to ignore specific linting rules for test files while keeping those rules enabled for the rest of the project.

When running `ruff check` from the command line, everything works as expected, the rules are ignored correctly for the test files.

However, when configuring Ruff through Zed's settings, it seems the per-file-ignores section is not applied.
Is it possible to achieve this setup using the Ruff Zed extension?

That setup works with command-line ruff
```pyproject.toml
[tool.ruff.lint.per-file-ignores]
"**/tests/**" = [
  "S101",   # assertions 
 ]
 ```
 
 That setup expected to work in Zed, but it doesn't.
 ```jsonc settings.json
{
  "lsp": {
    "ruff": {
      "initialization_options": {
        "settings": {
          "lint": {
            "select": [
              "ALL"
            ],
            "ignore": [
              "COM812", "COM819", "CPY"
            ],
            "per-file-ignores": {
              "**tests/**": [
                "S101"
              ]
            }
          }
        }
      }
    }
  }
}
```

**Comment by dhruvmanila** [comment]
In the editor context, only a finite number of settings are available which can be found at: https://docs.astral.sh/ruff/editors/settings/. We have discussed this previous on whether to allow specify all possible settings but there are some challenges there.

There are multiple solutions which you can utilize based on your setup:
1. Add a config file (`pyproject.toml` / `ruff.toml` / `.ruff.toml`) with the settings which is present in your workspace and it should be picked up by the server automatically and will be merged with the settings that's included in the `initialization_options`
2. If you want a central configuration i.e., apply these settings for all projects that you work on, you can create a Ruff config file and include the path to that file via [`ruff.configuration`](https://docs.astral.sh/ruff/editors/settings/#configuration).

Do you think either of the solutions mentioned above resolve your use-case? If not, can you say more about your setup?

**Reply by prepin** [reply]
Yeah, thanks! I see now that not every parameter can be set via editor config.
The issue was that Ruff, for some reason, ignores local `pyproject.toml` if there is no global configuration file. But if it is set, Ruff reads settings both from global ruff.toml and from local pyproject.toml.

So for those who might be reading this discussion trying to resolve the same issue as I found, here is what I've done:
- Created `ruff.toml` file at `~/.config/ruff/ruff.toml` and put most of my settings here.
- Added `"configuration": "~/.config/ruff/ruff.toml"` to Zed's `settings.json`.
- Added `extend = "~/.config/ruff/ruff.toml"` to local `pyproject.toml`, as Ruff does not perform a search for global ruff.toml if local file exists.

The only caveat I found is that if `per-file-ignores` is set only in global `ruff.toml` file, then `ruff check` calculates the path according to global `ruff.toml` file location, not according to project root. Which seems kind of strange.

```sh
ruff check --show-settings
...
linter.project_root = "/Users/prepin/dev/veles"
...
linter.per_file_ignores = {
    absolute_matcher = "/Users/prepin/.config/ruff/**/tests/**"
    basename_matcher = "**/tests/**"
...
```
          
Passing path to config file directly fixes that behavior.

```sh
ruff check --show-settings --config ~/.config/ruff/ruff.toml
...
linter.project_root = "/Users/prepin/dev/veles"
...
linter.per_file_ignores = {
    absolute_matcher = "/Users/prepin/dev/veles/**/tests/**"
    basename_matcher = "**/tests/**"
```

**Reply by dhruvmanila** [reply, marked as answer]
> The issue was that Ruff, for some reason, ignores local `pyproject.toml` if there is no global configuration file.

I don't think this should happen. If you have a local config file _and_ additional settings in the initialization options, the server should pick up both of them and merge unless you've overridden the [`ruff.configurationPreference`](https://docs.astral.sh/ruff/editors/settings/#configurationpreference) setting which I think you haven't.

Here's what I'm testing it against - I've a project whose root contains a `pyproject.toml` and a Python file with the following code:
```py
import math
import click
```

The Zed settings:
```json
{
  "lsp": {
    "ruff": {
      "initialization_options": {
        "settings": {
          "lint": {
            "extendSelect": ["I"]
          }
        }
      }
    }
  }
}
```

And, the `pyproject.toml` file contains:
```toml
[tool.ruff.lint]
ignore = ["F401"]
```

For the above file content, I'm getting `I` (organize imports) and no unused imports diagnostics.

Can you say more about your setup?

**Reply by prepin** [reply]
I've set up everything from scratch once again, and I can confirm that my initial guess was wrong. Ruff do respect values in `pyproject.toml` even if there is no common `ruff.toml` config present.

But I can reproduce that [extendIgnore](https://docs.astral.sh/ruff/editors/settings/#extendignore) initialization option is ignored by LSP version (see setup case # 4 and 5)

---
# Environment
## Software versions
* **macOS** v15.1.1 (24B91) arm64
* **Zed** v0.163.2
* **Zed Ruff extension** v0.1.0
* **Ruff** v0.8.0

Ruff installed via Homebrew, before installing Zed extension, so it uses brew-provided Zed instead of bundled one.

## Test file

```python
import math
import click


def hello_world(
    input: str,
) -> str:  # And this is long-long-long comment which violates linelength option. It should trigger rule E501.
    return "Hello World!"
```

# Setup cases

## # 1, without global config file

### Setup

**pyproject.toml**
```toml
not exist
```

**~/.config/ruff/ruff.toml**
```toml
not exist
```

**zed settings.json**
```json
{
  "languages": {
    "Python": {
      "language_servers": ["pyright", "ruff"],
      "format_on_save": "on",
      "tab_size": 4,
      "formatter": [
        {
          "code_actions": {
            "source.organizeImports.ruff": true,
            "source.fixAll.ruff": true
          }
        },
        {
          "language_server": {
            "name": "ruff"
          }
        }
      ]
    }
  },
  "lsp": {
    "ruff": {
      "initialization_options": {
        "settings": {
          "lineLength": 80,
          "lint": {}
        }
      }
    }
  }
}
```
### Result
| Ruff version | Output |
|--------|--------|
|Ruff via LSP:| F401|
|ruff check| F401 |

### Observation
Ruff founds problems using default rulesets.

## # 2, enable I and E rulesets in initialization options

### Setup

**pyproject.toml**
```toml
not exist
```

**~/.config/ruff/ruff.toml**
```toml
not exist
```

**zed settings.json**
```json
{
  "languages": {
    "Python": {
      "language_servers": ["pyright", "ruff"],
      "format_on_save": "on",
      "tab_size": 4,
      "formatter": [
        {
          "code_actions": {
            "source.organizeImports.ruff": true,
            "source.fixAll.ruff": true
          }
        },
        {
          "language_server": {
            "name": "ruff"
          }
        }
      ]
    }
  },
  "lsp": {
    "ruff": {
      "initialization_options": {
        "settings": {
          "lineLength": 80,
          "lint": {
            "extendSelect": ["I", "E"]
          }
        }
      }
    }
  }
}
```
### Result
| Ruff version | Detected  |
|--------|--------|
|Ruff via LSP:| I001, F401, E501|
|ruff check| F401 |

### Observation
LSP version of Ruff reports newly enabled rules.
Command line version does not, since it is not aware about editor config (which is expected).

## # 3, move rulesets to pyproject.toml

### Setup

**pyproject.toml**
```toml
[project]
name = "ruff-test"
version = "0.0.0"
requires-python = ">=3.13"

[tool.ruff]
line-length = 20
indent-width = 4

[tool.ruff.lint]
extend-select = ["I", "E"]
```

**~/.config/ruff/ruff.toml**
```toml
not exist
```

**zed settings.json**
```json
{
  "languages": {
    "Python": {
      "language_servers": ["pyright", "ruff"],
      "format_on_save": "on",
      "tab_size": 4,
      "formatter": [
        {
          "code_actions": {
            "source.organizeImports.ruff": true,
            "source.fixAll.ruff": true
          }
        },
        {
          "language_server": {
            "name": "ruff"
          }
        }
      ]
    }
  },
  "lsp": {
    "ruff": {
      "initialization_options": {
        "settings": {
          "lineLength": 80,
          "lint": { }
        }
      }
    }
  }
}
```
### Result
| Ruff version | Detected  |
|--------|--------|
|Ruff via LSP:| I001, F401, E501|
|ruff check| I001, F401, E501 |

### Observation
LSP version of Ruff reports newly enabled rules using config from pyproject.toml. LSP version of Ruff prefers line-length value from initialization options over one defined in pyproject.toml.
Command line version reports newly enabled rules using config from pyproject.toml

## # 4, set extendIgnore in initialization options

### Setup

**pyproject.toml**
```toml
[project]
name = "ruff-test"
version = "0.0.0"
requires-python = ">=3.13"

[tool.ruff]
line-length = 20
indent-width = 4

[tool.ruff.lint]
extend-select = ["I", "E"]
```

**~/.config/ruff/ruff.toml**
```toml
not exist
```

**zed settings.json**
```json
{
  "languages": {
    "Python": {
      "language_servers": ["pyright", "ruff"],
      "format_on_save": "on",
      "tab_size": 4,
      "formatter": [
        {
          "code_actions": {
            "source.organizeImports.ruff": true,
            "source.fixAll.ruff": true
          }
        },
        {
          "language_server": {
            "name": "ruff"
          }
        }
      ]
    }
  },
  "lsp": {
    "ruff": {
      "initialization_options": {
        "settings": {
          "lineLength": 80,
          "lint": {
            "extendIgnore": ["E501"]
          }
        }
      }
    }
  }
}
```
### Result
| Ruff version | Detected  |
|--------|--------|
|Ruff via LSP:| I001, F401, E501|
|ruff check| I001, F401, E501 |

### Observation
LSP version of Ruff ignores values set for [extendIgnore](https://docs.astral.sh/ruff/editors/settings/#extendignore) parameter in initialization options.

## # 5, move extendIgnore to pyproject.toml

### Setup

**pyproject.toml**
```toml
[project]
name = "ruff-test"
version = "0.0.0"
requires-python = ">=3.13"

[tool.ruff]
line-length = 20
indent-width = 4

[tool.ruff.lint]
extend-select = ["I", "E"]
extend-ignore = ["E501"]
```

**~/.config/ruff/ruff.toml**
```toml
not exist
```

**zed settings.json**
```json
{
  "languages": {
    "Python": {
      "language_servers": ["pyright", "ruff"],
      "format_on_save": "on",
      "tab_size": 4,
      "formatter": [
        {
          "code_actions": {
            "source.organizeImports.ruff": true,
            "source.fixAll.ruff": true
          }
        },
        {
          "language_server": {
            "name": "ruff"
          }
        }
      ]
    }
  },
  "lsp": {
    "ruff": {
      "initialization_options": {
        "settings": {
          "lineLength": 80,
          "lint": {}
        }
      }
    }
  }
}
```
### Result
| Ruff version | Detected  |
|--------|--------|
|Ruff via LSP:| I001, F401|
|ruff check| I001, F401 |

### Observation
LSP version of Ruff respects `extend-ignore` values set in `pyproject.toml`.

Command line version respects `extend-ignore` values set in `pyproject.toml`.

**Reply by dhruvmanila** [reply]
Wow, this is great! Thank you for taking the time to run these scenarios and providing all the details. I think the (4) case is a bug related to `lint.extendIgnore`, I've opened an issue for it #14665.

**Reply by dhruvmanila** [reply]
Actually, the problem is that the `extend-ignore` setting is deprecated which is why I think the `lint.extendIgnore` is not being considered. Can you try using the `lint.ignore` setting which is equivalent? It works on my end.

**Reply by prepin** [reply]
`lint.ignore` works for me as well

## # 6, replace extendIgnore with ignore

### Setup

**pyproject.toml**
```toml
[project]
name = "ruff-test"
version = "0.0.0"
requires-python = ">=3.13"

[tool.ruff]
line-length = 20
indent-width = 4

[tool.ruff.lint]
extend-select = ["I", "E"]
```

**~/.config/ruff/ruff.toml**
```toml
not exist
```

**zed settings.json**
```json
{
  "languages": {
    "Python": {
      "language_servers": ["pyright", "ruff"],
      "format_on_save": "on",
      "tab_size": 4,
      "formatter": [
        {
          "code_actions": {
            "source.organizeImports.ruff": true,
            "source.fixAll.ruff": true
          }
        },
        {
          "language_server": {
            "name": "ruff"
          }
        }
      ]
    }
  },
  "lsp": {
    "ruff": {
      "initialization_options": {
        "settings": {
          "lineLength": 80,
          "lint": {
            "ignore": ["E501"]
          }
        }
      }
    }
  }
}
```

| Ruff version | Detected  |
|--------|--------|
|Ruff via LSP:| I001, F401|
|ruff check| I001, F401, E501 |
